util/network-devp2p/src/discovery.rs

```diff
@@ -855,24 +869,70 @@ fn append_expiration(rlp: &mut RlpStream) {
 	rlp.append(&timestamp);
 }
 
-fn assemble_packet(packet_id: u8, payload: Bytes, secret: &Secret) -> Result<Bytes, Error> {
-	let mut packet = Bytes::with_capacity(payload.len() + 32 + 65 + 1);
-	packet.resize(32 + 65, 0); // Filled in below
-	packet.push(packet_id);
+
+/// Helper function to assemble node discovery packets
+///
+/// The packet format is: `hash | signature | payload | packet_type`, where the maximum packet length is 1280 bytes
+fn assemble_packet(packet_id: PacketKind, payload: Bytes, secret: &Secret) -> Result<(Bytes, H256), Error> {
+	let packet_len = payload.len() + HASH_LEN + SIGNATURE_LEN + PACKET_TYPE_LEN;
+
+	if !packet_has_valid_length(packet_len) {
+		warn!(target: "discovery", "Invalid packet length: {}", packet_len);
+		return Err(Error::BadProtocol);
+	}
+
+	let mut packet = Bytes::with_capacity(packet_len);
+	packet.resize(HASH_LEN + SIGNATURE_LEN, 0); // Filled in below
+	packet.push(packet_id as u8);
 	packet.extend(payload);
 
-	let hash = keccak(&packet[(32 + 65)..]);
-	let signature = match sign(secret, &hash) {
-		Ok(s) => s,
-		Err(e) => {
-			warn!(target: "discovery", "Error signing UDP packet");
-			return Err(Error::from(e));
-		}
-	};
-	packet[32..(32 + 65)].copy_from_slice(&signature[..]);
-	let signed_hash = keccak(&packet[32..]);
-	packet[0..32].copy_from_slice(signed_hash.as_bytes());
-	Ok(packet)
+	let signature = sign_payload_with_packet_id(secret, &packet[HASH_LEN + SIGNATURE_LEN ..])?;
+	packet[HASH_LEN..(HASH_LEN + SIGNATURE_LEN)].copy_from_slice(&signature[..]);
+	let signed_hash = keccak(&packet[HASH_LEN..]);
+	packet[..HASH_LEN].copy_from_slice(signed_hash.as_bytes());
+	Ok((packet, signed_hash))
+}
+
+/// Helper to disassemble node discovery packets
+fn disassemble_packet(packet: &[u8]) -> Result<(NodeId, &[u8], PacketKind, H256), Error> {
+	if !packet_has_valid_length(packet.len()) {
+		warn!(target: "discovery", "Invalid packet length: {}", packet.len());
+		return Err(Error::BadProtocol);
+	}
+
+	let payload_with_packet_id = &packet[HASH_LEN + SIGNATURE_LEN..];
+	let packet_kind = payload_with_packet_id[0];
+	let packet_kind: PacketKind = packet_kind.try_into().map_err(|e| {
```

Comment:
do this check earlier to avoid hashing, ecc recover when we get invalid/unsupported packet kinds